### ISSUE TITLE
Support for F# option types

### ISSUE THREAD

**Issue by rdavisau** [antmicro/Migrant#91]
Hi there, 

I noticed when testing FSI support added by #88 (thankyou!) that I was unable to serialize an F# option type. This seems to be the case both in FSI and when running as an ordinary console application. 

```fsharp
open Antmicro.Migrant
open System.IO

[<EntryPoint>]
let main argv = 

    let serializer = new Serializer();

    let option = Some "hello world"
    serializer.Serialize(option, new MemoryStream());

    0
```

gives:
```
System.Security.VerificationException: Operation could destabilize the runtime.
   at Write_FSharpOption`1(ObjectWriter , PrimitiveWriter , Object )
   at Antmicro.Migrant.ObjectWriter.WriteObjectInner(Object o)
   at Antmicro.Migrant.ObjectWriter.InvokeCallbacksAndWriteObject(Object o)
   at Antmicro.Migrant.ObjectWriter.WriteObject(Object o)
   at Antmicro.Migrant.Serializer.Serialize(Object obj, Stream stream)
```
on the `Serialize` call. 

I have produced on x64 .NET 4.5.2 and 4.6 with F# 4.0 using a build as at  47086047b56efb0986fbea749c4ba9549a9e8d5b.

**Comment by mateusz-holenko** [comment]
We recreated the issue and analyzed it, but still have no *good* solution. As for now I suggest to install a surogate for `FSharpOption` class; in C# it looks like:

```csharp
serializer.ForObject<FSharpOption<string>>()
    .SetSurrogate(x => new S { Field = x.Value });

serializer.ForSurrogate<S>()
    .SetObject(x => new FSharpOption<string>(x.Field));
```
```csharp
public class S
{
    public string Field { get; set; }
}
```
Unfortunately there is no simple way of installing generic surrogates right now. As a result, you will have to add each one manually, i.e., for `ints`, `floats`, `strings` etc.

We think about adding special `API` allowing to set surrogates for generic types, but this requires some architectural changes.

But returning to the core problem, it looks like setting `Value` field in `FSharpOption` class from generated IL is not allowed. We do not understand it, as it works for other similar classes and when using `Reflection` mode. We already asked question on *StackOverflow* hoping for some explanation: http://stackoverflow.com/q/31990025/980025

**Comment by rdavisau** [comment]
Thanks for looking into it. Very strange. I have linked the StackOverflow question on the F# Slack channel - with some luck it gets a little more attention!

**Comment by rdavisau** [comment]
I don't know if it helps, but I noticed you get the same error if you attempt to serialize an `FSharpList`:

```fsharp
open Antmicro.Migrant
open System.IO

[<EntryPoint>]
let main argv = 

    let serializer = new Serializer();

    let values = [1..5]
    serializer.Serialize(values, new MemoryStream());

    0
```
gives:
```fsharp

System.Security.VerificationException: Operation could destabilize the runtime.
>    at Write_FSharpList`1(ObjectWriter , PrimitiveWriter , Object )
   at Antmicro.Migrant.ObjectWriter.WriteObjectInner(Object o)
   at Antmicro.Migrant.ObjectWriter.InvokeCallbacksAndWriteObject(Object o)
   at Antmicro.Migrant.ObjectWriter.WriteObject(Object o)
   at Antmicro.Migrant.Serializer.Serialize(Object obj, Stream stream)
   at FSI_0037.main[a](a argv) in C:\Users\rdavis\Documents\Visual Studio 2015\Projects\ConsoleApplication2\ConsoleApplication3\Program.fs:line 9
   at <StartupCode$FSI_0037>.$FSI_0037.main@() in C:\Users\rdavis\Documents\Visual Studio 2015\Projects\ConsoleApplication2\ConsoleApplication3\Program.fs:line 13
Stopped due to error
```

**Comment by konrad-kruczynski** [comment]
Since to that time we haven't found a reason for verification fail nor the stackoverflow.com have given any answer and I've merged generic surrogates feature (#95), I recommend to use and am closing this ticket for now.
